proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+      [this PR](https://github.com/vmware-tanzu/pinniped/pull/1857).
+    - It should take care to correctly handle unicode characters that might exist in flag values when
+      invoking the subprocess, e.g. emojis in the upstream IDP name flag.
+      As long as the subprocess is successful, its stdout results can be ignored, because we only care about
+      what it cached in the session cache as a side effect.
+4. Now that the user is authenticated, the command will need to be able to construct a session cache key
+   exactly the same way that the "pinniped login oidc" subprocess did it, so it can be sure that it will be
+   looking up the same session in the cache.
+5. Next, it could open the session cache and read the access token from it. This access token should still be
+   valid/non-expired because of the "pinniped login oidc" subprocess that just succeeded.
+6. Finally, it could call the new Supervisor API to add consent for the PAT by sending the access token
+   for authentication and sending the SHA-256 hash of the PAT to add it to the allow list for the current user
+   identified by the access token. Revoking consent for a specific PAT would be a similar request,
+   and revoking consent for all PATs associated with the current user could work in a similar way
+   (but without needing to submit any hashed value of a specific PAT).
+
+#### Upgrades
+
+All changes will be backwards-compatible additions.
+
+#### Tests
+
+All aspects of the new features will be unit tested. Appropriate integration tests will also be added.
+
+#### New Dependencies
+
+We may like to use Google's golang client package for GitHub to help us call the GitHub API, although this would not
+be strictly necessary since these are simple REST API calls. It may help us implement things like pagination.
+
+#### Performance Considerations
+
+None.
+
+#### Observability Considerations
+
+Follow our pre-existing standards for error messages, log messages, custom resource status, etc.
+
+#### Security Considerations
+
+Some GitHub-specific considerations are already discussed above. Otherwise, this fits into the existing design
+of Pinniped without changing any of its existing security considerations.
+
+#### Usability Considerations
+
+No significant changes for end users in terms of how they authenticate or use kubectl.
+
+#### Documentation Considerations
+
+We will add API docs for GitHubIdentityProvider, and add docs on the website for authenticating with GitHub.
+
+### Other Approaches Considered
+
+None.
+
+## Open Questions
+
+- Needs investigation: How does the GitHub teams API deal with membership in nested teams? Does it flatten them for us?
+- Needs investigation: Does the GitHub API return org names with capital letters?
+  e.g. how does it return this org name? https://github.com/Broadcom (note the capital "B").
+  Should we treat the configured org names as case-sensitive?
```

Comment:
I think we should treat configured org names as case-sensitive.

We can attempt to validate the provided org names and provide the user with a status condition indicating that they have used the wrong casing for their org name and even provide the correct casing inside the status condition as a suggestion.

Note that org information is never private (https://github.com/orgs/community/discussions/23157#discussioncomment-3239405) so Pinniped can always query the org information to obtain the correct casing.

Reply:
Pinniped could even use the provided client id/secret to make authenticated requests as the client, to avoid the severely low 60 requests/hour for unauthenticated API calls.

https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28

Reply:
On the other hand, it looks like github org names are just case-insensitive.

See https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#get-an-organization--fine-grained-access-tokens, which says
> The organization name. The name is not case sensitive.

Vault treats github org names as case-insensitive: https://github.com/hashicorp/vault/pull/1362
crates.io treats github org names as case-insensitive: https://github.com/rust-lang/crates.io/pull/1171

Note that authenticating as an app is extremely complicated in GitHub.

Thanks @cfryanr for the following: 
In order to auth as the GitHub App, you need to create a JWT. To create the JWT, you need the private key that should be used to sign the JWT. To have a private key, you must go to the UI and create a private key for the App. It will only show you the private key once, so you need to save that private key somewhere. So for Pinniped to make an API call as the App, it is not enough to know the App’s client ID and client secret… Pinniped would also need to know the App’s private key, which the user would need to configure.

- https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app
- https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app
- https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/managing-private-keys-for-github-apps

Reply:
Note that when constructing the list of upstream groups, which look like `org-name:team-name-or-slug`, we should preserve the case of the organization's `login` attribute and the case of the team-name-or-slug, both of which should be retrieved from the `/user/teams` endpoint https://docs.github.com/en/rest/teams/teams?apiVersion=2022-11-28#list-teams-for-the-authenticated-user.

Reply:
TODO: Update proposal to say that org membership comparisons shall be done as case-insensitive string comparisons, but case shall be preserved in the org names when they appear in downstream group names. Kubernetes groups as case sensitive, and CEL identity transformations can optionally be used to change their case if desired.